src/bsoncxx/array/view.cpp

```diff
@@ -139,6 +140,19 @@ element view::operator[](std::uint32_t i) const {
     return *(this->find(i));
 }
 
+view::const_iterator view::find_value(bsoncxx::types::bson_value::view v) const {
```

Comment:
I thought that might be confusing to users because the two methods work so differently.  The existing find() method looks up an element at the given index--which is not what I expected it to do when I first looked at this class!

Really, I think it would be more clear if the existing `array::find()` method were renamed to `array::at()`, to match the API of `std::array`.  Then we would be free to repurpose `array::find()` to act like`array::find_value()`, which would then also behave more like `std::find` and maybe be more clear to users.

If you and @kevinAlbs agree with all that, then here's what I think we could do:
- rename the existing `array::find()` to `array::at()` and deprecate `array::find()`
- implement `array::find_value()` as shown here and maybe add `array::find_value_if()`
- open a JIRA ticket to deprecate `array::find_value()` and rename it to `array::find()` in the next release (same for `find_value_if`)

Reply:
> rename the existing array::find() to array::at() and deprecate array::find()

> open a JIRA ticket to deprecate array::find_value() and rename it to array::find() in the next release (same for find_value_if)


I agree with `at` being a better name than `find`. I am in favor of adding `at`.

But I am opposed to deprecating and repurposing `find`.

Given that you can construct a `bson_value` from an int32, that might silently change behavior when a user upgrades.

Suppose a user has this code, intending to look up element at position 3:
```
auto iter = arr.find(3);
```
Then they upgrade to C++ 3.8, and we've change `find` to mean "find by searching bson_value", I think this would still compile, but have a different meaning.

> implement array::find_value() as shown here and maybe add array::find_value_if()

`find_value` seems like a clear enough name to me. As does `search`.

If we want to be consistent with `std::array`, it does not provide `find` or `find_if`. Instead, users use `std::find` and `std::find_if` with iterators. We could take the same approach. If we made a `bsoncxx::element` comparable to `bsoncxx::bson_value`, I think users could use `std::find` and `std::find_if`.

That being said `arr.find_value("test")` is more concise than `std::find (arr.begin(), arr.end(), "test"))`. `find_value` and `has_value` seem like reasonable conveniences. But I think we should draw the line somewhere too.

Reply:
Perhaps we should rename `find()` to `at()` (in a separate ticket) and leave this as `find_value()` then?

Reply:
SGTM if @bazile-clyde agrees.

Reply:
> I thought that might be confusing to users because the two methods work so differently. The existing find() method looks up an element at the given index--which is not what I expected it to do when I first looked at this class!

Ah, that makes sense. It’d be really confusing to have a subtle difference between finding by index or value.

> Really, I think it would be more clear if the existing array::find() method were renamed to array::at(), to match the API of std::array.

I think so too. We could (and probably should for consistency with `std::array`) also add bounds checking for `at`. So `throw` instead of returning the end iterator. Especially since we already have a non-throwing "find" with `operator[]`.

> Then we would be free to repurpose array::find() to act like array::find_value(), which would then also behave more like std::find and maybe be more clear to users.

That being said, I’m not sure why we need `find` or `find_value` though. Sequence containers in the STL don’t have `find` methods and instead rely on `std::find` to look for specific elements by value. The `find` methods for associative containers are used to look up elements by keys, not values. 

As is, `find_value` behaves identically to `std::find`. The only difference is how we compare the elements. Why not just create an `operator==` that compares array elements to values? Then users can use `std::find` without a custom compare function, which is essentially what `find_value` is abstracting away.

Reply:
I think that adding `find_value` is a nice shortcut, and a little bit nicer to use than `std::find` for an operation that I imagine being somewhat common for users.  Especially considering that it might be a little confusing to users that `std::find` would compare `element` and `value` and not two things of the same type.  But, it does differ from the API of an STL container, and if you and @kevinAlbs agree that it would be better to encourage users to use `std::find` for element lookup, I'll change the test to use `std::find` and add public `element` to `value` comparison functions instead.

I've opened a separate ticket to rename `find()` to `at()`.

Reply:
I discussed this with @bazile-clyde. Though I do see the convenience of `find_value`, I am in favor of removing `find_value` in and adding an `operator==` overload for `bson_value` and `element` for consistency with STL.

Reply:
Fair enough! Done.